### PR TITLE
move getMaxVols function to predicates.go and add some NewVolumeCount…

### PR DESCRIPTION
**Pull request by jiulongzaitian**
…Predicate funcs

Signed-off-by: zhangjie <zhangjie0619@yeah.net>

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
move getMaxVols function to predicates.go and add some NewVolumeCountPredicate funcs
```

### CODE REVIEW

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults_test.go [hunk 1]
```diff
@@ -15,48 +15,3 @@ limitations under the License.
 */
 
 package defaults
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults_test.go]: No tests in this file. Better delete it?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -43,6 +44,19 @@ import (
 	"k8s.io/metrics/pkg/client/clientset_generated/clientset"
 
 	"github.com/golang/glog"
+	"os"
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: No `gofmt`?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -177,14 +191,24 @@ type VolumeFilter struct {
 	FilterPersistentVolume func(pv *v1.PersistentVolume) (id string, relevant bool)
 }
 
+func NewMaxEBSVolumeCountPredicate(pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: I still prefer the old `NewMaxPDVolumeCountPredicate`.

Actually there is no need to add three new-added methods, which have the same exact apis.

File: plugin/pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 4]
```diff
@@ -30,6 +30,7 @@ import (
 	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
 	schedulertesting "k8s.io/kubernetes/plugin/pkg/scheduler/testing"
 	schedutil "k8s.io/kubernetes/plugin/pkg/scheduler/util"
+	"os"
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates_test.go]: No `gofmt`?

**Review by dixudx — CHANGES_REQUESTED** [review]
Since `NewMaxPDVolumeCountPredicate` is only called by these three "*VolumeCount", I think it is safe to refactor it by adding a new param to incorporate these three cases. I'd prefer this way.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Similar to what @dixudx said, I wonder why we need to wrap `newMaxPDVolumeCountPredicate`?

Comment (jiulongzaitian) [reply plugin/pkg/scheduler/algorithm/predicates/predicates.go]: @dixudx  @bsalamat  Thank you for your suggestion。 I add a new param to incorporate these three cases. please view again.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 5]
```diff
@@ -178,13 +196,30 @@ type VolumeFilter struct {
...
 // NewMaxPDVolumeCountPredicate creates a predicate which evaluates whether a pod can fit based on the
-// number of volumes which match a filter that it requests, and those that are already present.  The
-// maximum number is configurable to accommodate different systems.
+// number of volumes which match a filter that it requests, and those that are already present.
 //
 // The predicate looks for both volumes used directly, as well as PVC volumes that are backed by relevant volume
 // types, counts the number of unique volumes, and rejects the new pod if it would place the total count over
 // the maximum.
-func NewMaxPDVolumeCountPredicate(filter VolumeFilter, maxVolumes int, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
+func NewMaxPDVolumeCountPredicate(filterName string, pvInfo PersistentVolumeInfo, pvcInfo PersistentVolumeClaimInfo) algorithm.FitPredicate {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Are you sure to remove parameter `maxVolumes int`?

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -195,6 +230,21 @@ func NewMaxPDVolumeCountPredicate(filter VolumeFilter, maxVolumes int, pvInfo Pe
 	return c.predicate
 }
 
+// getMaxVols checks the max PD volumes environment variable, otherwise returning a default value
+func getMaxVols(defaultVal int) int {
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: This may not what we expected. The maximum volumes here can only be set by `os.env`. This is not the k8s way. I think we'd keep the old configurable way. @bsalamat WDYT?

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go [hunk 7]
```diff
@@ -34,19 +30,11 @@ import (
...
 
 const (
-	// DefaultMaxGCEPDVolumes defines the maximum number of PD Volumes for GCE
-	// GCE instances can have up to 16 PD volumes attached.
-	DefaultMaxGCEPDVolumes = 16
-	// DefaultMaxAzureDiskVolumes defines the maximum number of PD Volumes for Azure
-	// Larger Azure VMs can actually have much more disks attached.
-	// TODO We should determine the max based on VM size
-	DefaultMaxAzureDiskVolumes = 16
+
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: Remove this new line.

File: plugin/pkg/scheduler/factory/plugins.go [hunk 8]
```diff
@@ -118,6 +118,9 @@ func RegisterMandatoryFitPredicate(name string, predicate algorithm.FitPredicate
 func RegisterFitPredicateFactory(name string, predicateFactory FitPredicateFactory) string {
 	schedulerFactoryMutex.Lock()
 	defer schedulerFactoryMutex.Unlock()
+	if predicateFactory == nil {
```
Comment (dixudx) [inline plugin/pkg/scheduler/factory/plugins.go]: What would happen if missing this nil pointer checking?

I think we should not introduce unnecessary changes if everything works well.

Comment (dixudx) [reply plugin/pkg/scheduler/factory/plugins.go]: Also I think the checking of `predicateFactory` should be done in the upper calling method.

File: plugin/pkg/scheduler/algorithm/predicates/predicates.go [hunk 9]
```diff
@@ -45,6 +47,22 @@ import (
...
+	// GCE instances can have up to 16 PD volumes attached.
+	DefaultMaxGCEPDVolumes = 16
+	// DefaultMaxAzureDiskVolumes defines the maximum number of PD Volumes for Azure
+	// Larger Azure VMs can actually have much more disks attached.
+	// TODO We should determine the max based on VM size
+	DefaultMaxAzureDiskVolumes = 16
+	// KubeMaxPDVols defines the maximum number of PD Volumes per kubelet
+	KubeMaxPDVols = "KUBE_MAX_PD_VOLS"
+
+	EBSVolume       = "EBSVolume"
```
Comment (dixudx) [inline plugin/pkg/scheduler/algorithm/predicates/predicates.go]: Adding some comments here on the const value?

**Review by dixudx — CHANGES_REQUESTED** [review]
Don't change the api if unnecessary. Especially removing configurable parameters.

File: plugin/pkg/scheduler/factory/plugins.go (hunk 8 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/factory/plugins.go]: +1 to @dixudx. The addition of this check is not necessary in this PR.

**Review by bsalamat — COMMENTED** [review]
Thanks, @jiulongzaitian! Looks good in general. I have a minor comment.

**Comment by jiulongzaitian** [comment]
@dixudx  @bsalamat  thank you very much for your suggest, please check new code if you have time.

File: plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go (hunk 7 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/algorithmprovider/defaults/defaults.go]: Please remove this line. The rest LGTM.

**Comment by bsalamat** [comment]
@jiulongzaitian Did you rebase your changes incorrectly? I no longer see all the changes. There is only one deleted test.

**Comment by jiulongzaitian** [comment]
hi  @bsalamat , sorry , i rebased incorrectly. please view again when you have time. thank you .

**Comment by bsalamat** [comment]
@jiulongzaitian Sorry, I was on vacation last week and didn't get the chance to look at this.
Why did you close the PR?

**Comment by jiulongzaitian** [comment]
hi @bsalamat ,thank you，test case have some errors , i am  sorry  to close this pr, i  reopen it now.

**Comment by jiulongzaitian** [comment]
hi  @bsalamat  , i remove maxVolumes parameter from   NewMaxPDVolumeCountPredicate function , 
 but TestEBSVolumeCountConflicts function need this parameter.   Do you think  is it necessary to delete this parameter ?

**Comment by bsalamat** [comment]
@jiulongzaitian 
I think the current code without the `maxVolume` parameter is cleaner than before. In order to fix the `TestEBSVolumeCountConflicts` test, you can keep `maxVols` in the test cases and inside the `for _, test := range tests` loop call:
os.Setenv(KubeMaxPDVols, strconv.FormatInt(test.maxVols))
This will set the environment variable and tests `getMaxVols` better.

**Comment by jiulongzaitian** [comment]
hi @bsalamat  Sorry, I was on vacation last week(China National Day), I will try it to follow your advice. thank you.

**Comment by bsalamat** [comment]
Thanks, @jiulongzaitian! LGTM.

Please run hakc/update-bazel.sh

**Comment by bsalamat** [comment]
@jiulongzaitian I think you need to run `hack/update-bazel.sh` in your work directory and commit the changes.

**Comment by jiulongzaitian** [comment]
hi  @bsalamat thank you for help !

**Comment by bsalamat** [comment]
@jiulongzaitian Thank you for doing the work!
